src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/FSharpCoreReflectionProxy.cs

```diff
@@ -0,0 +1,224 @@
+    /// Proxy class used to access FSharp.Core metadata and reflection APIs that are not statically available to System.Text.Json.
+    /// </summary>
+    internal sealed class FSharpCoreReflectionProxy
+    {
+        public const string FSharpCoreUnreferencedCodeMessage = "Uses Reflection to access FSharp.Core components, which requires unreferenced code.";
+
+        private static object s_lockObj = new object();
+        private static FSharpCoreReflectionProxy? s_singletonInstance;
+
+        // Every type generated by the F# compiler is annotated with the CompilationMappingAttribute
+        // containing all relevant metadata required to determine its kind:
+        // https://fsharp.github.io/fsharp-core-docs/reference/fsharp-core-compilationmappingattribute.html#SourceConstructFlags
+        private const string CompilationMappingAttributeTypeName = "Microsoft.FSharp.Core.CompilationMappingAttribute";
+        private readonly Type _compilationMappingAttributeType;
+        private readonly MethodInfo? _sourceConstructFlagsGetter;
+
+        private readonly Type? _fsharpOptionType;
+        private readonly Type? _fsharpListType;
+        private readonly Type? _fsharpSetType;
+        private readonly Type? _fsharpMapType;
+
+        private readonly MethodInfo? _fsharpListCtor;
+        private readonly MethodInfo? _fsharpSetCtor;
+        private readonly MethodInfo? _fsharpMapCtor;
+
+        /// <summary>
+        /// Checks if the provided System.Type instance is emitted by the F# compiler.
+        /// If true, also initializes the proxy singleton for future use by other F# types.
+        /// </summary>
+        public static bool IsFSharpType(Type type)
+        {
+            if (s_singletonInstance is null)
+            {
+                if (GetFSharpCoreAssembly(type) is Assembly fsharpCoreAssembly)
+                {
+                    // Type is F# type, initialize the singleton instance.
+                    lock (s_lockObj)
+                    {
+                        s_singletonInstance ??= new FSharpCoreReflectionProxy(fsharpCoreAssembly);
+                    }
+
+                    return true;
+                }
+
+                return false;
+            }
+
+            return s_singletonInstance.GetFSharpCompilationMappingAttribute(type) is not null;
+        }
+
+        /// <summary>
+        /// Gets the singleton proxy instance; prerequires a successful IsFSharpType call for proxy initialization.
+        /// </summary>
+        public static FSharpCoreReflectionProxy Instance
+        {
+            get
+            {
+                Debug.Assert(s_singletonInstance is not null, "should be initialized via a successful IsFSharpType call.");
+                return s_singletonInstance;
+            }
+        }
+
+        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2026:RequiresUnreferencedCode", Justification = "The public methods are marked RequiresUnreferencedCode.")]
```

Comment:
These are not valid justifications for suppressing the warnings. Instead, this method should be marked with `[RequiresUnreferencedCode]`. That way all callers of these methods (now and in the future) will get a warning.